lib/gui/app/pages/main/controllers/flash.js

```diff
@@ -25,19 +26,88 @@ const exceptionReporter = require('../../../modules/exception-reporter')
 const imageWriter = require('../../../modules/image-writer')
 const path = require('path')
 const store = require('../../../../../shared/store')
+const constraints = require('../../../../../shared/drive-constraints')
+const availableDrives = require('../../../../../shared/models/available-drives')
 
 module.exports = function (
+  $q,
   $state,
   $timeout,
-  FlashErrorModalService
+  FlashErrorModalService,
+  WarningModalService,
+  DriveSelectorService
 ) {
+  /**
+   * @summary Spawn a confirmation warning modal
+   * @function
+   * @public
+   *
+   * @param {Array<String>} warningMessages - warning messages
+   * @returns {Promise} warning modal promise
+   *
+   * @example
+   * confirmationWarningModal([ 'Hello, World!' ])
+   */
+  const confirmationWarningModal = (warningMessages) => {
+    return WarningModalService.display({
+      confirmationLabel: 'Continue',
+      rejectionLabel: 'Change',
+      description: [
+        warningMessages.join('\n\n'),
+        'Are you sure you want to continue?'
+      ].join(' ')
+    })
+  }
+
+  /**
+   * @summary Display warning tailored to the warning of the current drives-image pair
+   * @function
+   * @public
+   *
+   * @param {Array<Object>} drives - list of drive objects
+   * @param {Object} image - image object
+   * @returns {Promise}
+   *
+   * @example
+   * displayTailoredWarning(drives, image).then(() => {
+   *   console.log('Continue pressed')
+   * }).catch(() => {
+   *   console.log('Change pressed')
+   * })
+   */
+  const displayTailoredWarning = (drives, image) => {
+    const warningMessages = _.reduce(drives, (accumMessages, drive) => {
+      if (constraints.isDriveSizeLarge(drive)) {
+        return accumMessages.concat(messages.warning.largeDriveSize(drive))
+      } else if (!constraints.isDriveSizeRecommended(drive, image)) {
+        return accumMessages.concat(messages.warning.unrecommendedDriveSize(image, drive))
+      }
+
+      return accumMessages
+    }, [])
+
+    if (!warningMessages.length) {
+      // TODO(Shou): we should consider adding the same warning dialog for system drives and remove unsafe mode
+      return $q.resolve()
+    }
+
+    return confirmationWarningModal(warningMessages).then((value) => {
+      if (!value) {
+        DriveSelectorService.open()
+        return $q.reject()
+      }
+
+      return $q.resolve()
+    })
+  }
+
   /**
    * @summary Flash image to drives
    * @function
    * @public
    *
    * @param {Object} image - image
-   * @param {Array<Object>} drives - drives
+   * @param {Array<Object>} devices - drive devices
```

Comment:
The type doc here needs updating to `Array<String>` too I think

Reply:
Done that now